qiskit/extensions/unitary.py

```diff
@@ -153,16 +153,26 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
             QiskitError: Invalid ctrl_state.
             ExtensionError: Non-unitary controlled unitary.
         """
+        from qiskit.quantum_info.synthesis.qsd import (  # pylint: disable=cyclic-import
+            qs_decomposition,
+        )
+
         mat = self.to_matrix()
         cmat = _compute_control_matrix(mat, num_ctrl_qubits, ctrl_state=None)
-        iso = isometry.Isometry(cmat, 0, 0)
+        try:
+            cmat_def = qs_decomposition(cmat)
+        except QiskitError as qerr:
+            if "CS decomposition error" in repr(qerr):
+                from qiskit.extensions.quantum_initializer.isometry import Isometry
```

Comment:
It'd be better to define a custom subclass of `QiskitError` that you can catch specifically, like
```python
class QSDError(QiskitError):
   """Errors specifically related to the quantum Shannon decomposer failing."""
```

and then catch that.